tasks/rh-sign-image-cosign/rh-sign-image-cosign.yaml

```diff
@@ -73,28 +82,68 @@ spec:
             echo "No valid file was provided as signRegistryAccessPath."
             exit 1
         fi
+        PUBLIC_KEY_FILE=$(mktemp)
+        echo -n "$PUBLIC_KEY" > "$PUBLIC_KEY_FILE"
+        RUNNING_JOBS="\j" # Bash parameter for number of jobs currently running
 
         run_cosign () { # Expected arguments are [digest_reference, tag_reference]
-            # Upload transparency log when rekor url is specified
-            if [ -v REKOR_URL ]; then
-                COSIGN_COMMON_ARGS="-y --rekor-url=$REKOR_URL --key $SIGN_KEY"
-            else
-                COSIGN_COMMON_ARGS="--tlog-upload=false --key $SIGN_KEY"
-            fi
-            echo "Signing manifest $1 ($2)"
             attempt=0
+            backoff1=2
+            backoff2=3
             until [ "$attempt" -gt "$(params.retries)" ] ; do # 3 retries by default
-                cosign -t 3m0s sign\
-                ${COSIGN_COMMON_ARGS}\
-                --sign-container-identity "$2"\
-                "$1" && break
+                cosign "$@" && break
+                sleep $backoff2
+
+                # Fibbonaci backoff
+                old_backoff1=$backoff1
+                backoff1=$backoff2
+                backoff2=$((old_backoff1 + backoff2))
+
                 attempt=$((attempt+1))
             done
             if [ "$attempt" -gt "$(params.retries)" ] ; then
               echo "Max retries exceeded."
               exit 1
             fi
         }
+        function check_existing_signatures() {
+          local identity=$1
+          local reference=$2
+          local digest=$3
+          if [ -v REKOR_URL ]; then
+              COSIGN_REKOR_ARGS="--rekor-url=$REKOR_URL"
+          else
+              COSIGN_REKOR_ARGS="--insecure-ignore-tlog=true"
+          fi
+          verify_output=$(run_cosign verify "$COSIGN_REKOR_ARGS" "$reference")
```

Comment:
Are we sure this will not print anything to stdout? If so, it will spoil the result of this function. So maybe better to add `>&2` at the end of the command?

Reply:
I'm not sure I understand you. Stdout will be captured in verify_output

Reply:
Ah, sorry, maybe it was too early in the morning :) And it actually prints the json that you then parse, I didn't realize that. So that should be fine.

There is one potential issue that I've hit before though. Here you run `run_cosign` in a new shell (because it's in `$()`). If you run just one command this way, that's usually fine. But if you run your own function, inside that context `set -e` will no longer be set. I *think* it shouldn't be a problem here because if retries fail, you call `exit 1`, so it will exit the subshell immediately, but something to keep in mind.

Reply:
yeah, but running check_and_sign in background already runs in subshell

Reply:
If that was the case, then that would be a potential issue as well. But I don't think that's the case. Why do you think `check_and_sign` runs in a subshell? Just because it's a function? Functions don't run in subshell, at least according to this: https://stackoverflow.com/questions/36158686/does-a-shell-function-run-in-a-sub-shell

Reply:
I also wasn't sure what runs in subshell and what doesn't , so I asked chatgpt and it told me: Commands run in the background (using &) are executed in a subshell.

Reply:
OK, then maybe it is :) Anyway, I just wanted to point out that things that run in subshell don't have the same setting. In that case of a function, that can be fixed by adding `set -eux` to the beginning of the function. But I'm not saying this is needed right now, just something to keep in mind.